File: token-client-core/src/test/java/no/nav/security/token/support/client/core/TestUtils.java
package no.nav.security.token.support.client.core;

import com.nimbusds.jwt.JWT;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.PlainJWT;
import com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.*;
import java.util.function.Consumer;

import static com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod.CLIENT_SECRET_BASIC;
import static com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod.PRIVATE_KEY_JWT;
import static org.assertj.core.api.Assertions.assertThat;

@SuppressWarnings("WeakerAccess")
public class TestUtils {

    public static final String CONTENT_TYPE_FORM_URL_ENCODED = "application/x-www-form-urlencoded;charset=UTF-8";
    public static final String CONTENT_TYPE_JSON = "application/json;charset=UTF-8";

    public static ClientProperties clientProperties(String tokenEndpointUrl, OAuth2GrantType oAuth2GrantType) {
        return ClientProperties.builder()
            .grantType(oAuth2GrantType)
            .scope(List.of("scope1", "scope2"))
            .tokenEndpointUrl(URI.create(tokenEndpointUrl))
            .authentication(ClientAuthenticationProperties.builder("client1",CLIENT_SECRET_BASIC)
                .clientSecret("clientSecret1")
                .build())
            .build();
    }

    public static ClientProperties tokenExchangeClientProperties(
        String tokenEndpointUrl,
        OAuth2GrantType oAuth2GrantType,
        String clientPrivateKey
    ) {
        return ClientProperties.builder()
            .grantType(oAuth2GrantType)
            .tokenEndpointUrl(URI.create(tokenEndpointUrl))
            .authentication(ClientAuthenticationProperties.builder("client1",PRIVATE_KEY_JWT)
                .clientJwk(clientPrivateKey)
                .build())
            .tokenExchange(ClientProperties.TokenExchangeProperties.builder()
                .audience("audience1")
                .build())
            .build();
    }

    public static void withMockServer(Consumer<MockWebServer> test) throws IOException {
        MockWebServer server = new MockWebServer();
        server.start();
        test.accept(server);
        server.shutdown();
    }

    public static MockResponse jsonResponse(String json) {
        return new MockResponse()
            .setHeader("Content-Type", "application/json;charset=UTF-8")
            .setBody(json);
    }

    public static void assertPostMethodAndJsonHeaders(RecordedRequest recordedRequest) {
        assertThat(recordedRequest.getMethod()).isEqualTo("POST");
        assertThat(recordedRequest.getHeader("Accept")).isEqualTo(CONTENT_TYPE_JSON);
        assertThat(recordedRequest.getHeader("Content-Type")).isEqualTo(CONTENT_TYPE_FORM_URL_ENCODED);
    }

    public static String decodeBasicAuth(RecordedRequest recordedRequest) {
        return Optional.ofNullable(recordedRequest.getHeaders().get("Authorization"))
            .map(s -> s.split("Basic "))
            .filter(pair -> pair.length == 2)
            .map(pair -> Base64.getDecoder().decode(pair[1]))
            .map(bytes -> new String(bytes, StandardCharsets.UTF_8))
            .orElse("");
    }

    public static JWT jwt(String sub) {
        Instant expiry = LocalDateTime.now().atZone(ZoneId.systemDefault()).plusSeconds(60).toInstant();
        return new PlainJWT(new JWTClaimsSet.Builder()
            .subject(sub)
            .audience("thisapi")
            .issuer("someIssuer")
            .expirationTime(Date.from(expiry))
            .claim("jti", UUID.randomUUID().toString())
            .build());
    }

    public static String encodeValue(String value) {
        String encodedUrl = null;
        try {
            encodedUrl = URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return encodedUrl;
    }
}

File: token-client-core/src/test/kotlin/no/nav/security/token/support/client/core/TestUtils.kt
package no.nav.security.token.support.client.core

import com.nimbusds.common.contenttype.ContentType.APPLICATION_JSON
import com.nimbusds.common.contenttype.ContentType.APPLICATION_URLENCODED
import com.nimbusds.jwt.JWTClaimsSet.Builder
import com.nimbusds.jwt.PlainJWT
import com.nimbusds.oauth2.sdk.GrantType
import com.nimbusds.oauth2.sdk.auth.ClientAuthenticationMethod.CLIENT_SECRET_BASIC
import mockwebserver3.MockResponse
import mockwebserver3.MockWebServer
import mockwebserver3.RecordedRequest
import java.io.UnsupportedEncodingException
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime.now
import java.time.ZoneId.systemDefault
import java.util.*
import no.nav.security.token.support.client.core.ClientAuthenticationProperties.Companion.builder
import no.nav.security.token.support.client.core.ClientProperties.Companion.builder
import org.assertj.core.api.Assertions.assertThat

object TestUtils {
    @JvmStatic
    fun clientProperties(tokenEndpointUrl : String, oAuth2GrantType : GrantType) =
        builder(oAuth2GrantType, builder("client1", CLIENT_SECRET_BASIC)
            .clientSecret("clientSecret1")
            .build())
            .scope(listOf("scope1", "scope2"))
            .tokenEndpointUrl(URI.create(tokenEndpointUrl))
            .build()

    @JvmStatic
    fun withMockServer(test: (MockWebServer) -> Unit) {
        MockWebServer().apply {
            start()
            test(this)
            close()
        }
    }

    @JvmStatic
    fun jsonResponse(json : String) = MockResponse(body = json).apply {
        headers.newBuilder().add("Content-Type", "application/json").build()
    }

    @JvmStatic
    fun assertPostMethodAndJsonHeaders(recordedRequest : RecordedRequest) {
        assertThat(recordedRequest.method).isEqualTo("POST")
        assertThat(recordedRequest.headers.get("Accept")).isEqualTo("$APPLICATION_JSON")
        assertThat(recordedRequest.headers.get("Content-Type")).isEqualTo("$APPLICATION_URLENCODED")
    }

    @JvmStatic
    fun decodeBasicAuth(recordedRequest : RecordedRequest) : String {
        return Optional.ofNullable(recordedRequest.headers["Authorization"])
            .map { s : String -> s.split("Basic ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray() }
            .filter { pair : Array<String> -> pair.size == 2 }
            .map { pair : Array<String> -> Base64.getDecoder().decode(pair[1]) }
            .map { bytes : ByteArray? -> String(bytes!!, StandardCharsets.UTF_8) }
            .orElse("")
    }

    @JvmStatic
    fun jwt(sub : String?) = PlainJWT(Builder()
        .subject(sub)
        .audience("thisapi")
        .issuer("someIssuer")
        .expirationTime(Date.from(now().atZone(systemDefault()).plusSeconds(60).toInstant()))
        .claim("jti", UUID.randomUUID().toString())
        .build())

    @JvmStatic
    fun encodeValue(value : String?) : String? {
        var encodedUrl : String? = null
        try {
            encodedUrl = URLEncoder.encode(value, StandardCharsets.UTF_8.toString())
        }
        catch (e : UnsupportedEncodingException) {
            e.printStackTrace()
        }
        return encodedUrl
    }
}